presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/RuleStore.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+
+import java.util.List;
+import java.util.Set;
+
+public class RuleStore
+{
+    private final ListMultimap<Class<? extends PlanNode>, Rule> rulesByClass;
+    private final List<Rule> unclassifiedRules;
+
+    private RuleStore(ListMultimap<Class<? extends PlanNode>, Rule> rulesByClass, List<Rule> unclassifiedRules)
+    {
+        this.rulesByClass = ImmutableListMultimap.copyOf(rulesByClass);
+        this.unclassifiedRules = ImmutableList.copyOf(unclassifiedRules);
+    }
+
+    public List<Rule> getCandidates(PlanNode planNode)
+    {
+        return ImmutableList.<Rule>builder()
+                .addAll(rulesByClass.get(planNode.getClass()))
+                .addAll(unclassifiedRules)
+                .build();
+    }
+
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    public static class Builder
+    {
+        private final ImmutableListMultimap.Builder<Class<? extends PlanNode>, Rule> rulesByClass = ImmutableListMultimap.builder();
+        private final ImmutableList.Builder<Rule> nonPartitionedRules = ImmutableList.builder();
+
+        public Builder register(Set<Rule> newRules)
+        {
+            newRules.forEach(this::register);
+            return this;
+        }
+
+        public Builder register(Rule newRule)
+        {
+            Pattern pattern = newRule.getPattern();
+            if (pattern == Pattern.AnyNode.INSTANCE) {
+                nonPartitionedRules.add(newRule);
+            }
+            else if (pattern instanceof Pattern.MatchNodeClass) {
+                rulesByClass.put(((Pattern.MatchNodeClass) pattern).getNodeClass(), newRule);
```

Comment:
This neglects inheritance and implemented interfaces. All the extended classes / implemented interfaces of the pattern's `nodeClass` should be `put` into the `rulesByClass` multimap. Otherwise a rule specifying that it matches a `PlanNode` or a `SetOperationNode` will not ever be fired (will not be included in `getCandidates`)

We can't expect the users will only match on concrete classes. Not without a precondition check and javadoc. And the future dev's fury when they encounter this limitation.